geth/rpc/call_raw.go

```diff
@@ -104,12 +105,42 @@ func methodAndParamsFromBody(body string) (string, []interface{}, json.RawMessag
 	return msg.Method, params, msg.ID, nil
 }
 
+// unmarshalMessage tries to unmarshal JSON-RPC message.
+// somehow JSON-RPC input from web3.js can be in two forms:
+//
+// object:  {"jsonrpc":"2.0", …}
+// array:   [{"jsonrpc":"2.0", …}]
+//
+// unmarhsalMessage tries first option and in case of error,
+// tries to unmarshal it as an array.
+//
+// TODO(divan): fix the source of this error and cleanup.
 func unmarshalMessage(body string) (*jsonrpcMessage, error) {
 	var msg jsonrpcMessage
 	err := json.Unmarshal([]byte(body), &msg)
+	// check for array case
+	if e, ok := err.(*json.UnmarshalTypeError); ok {
+		if e.Value == "array" {
+			return unmarshalMessageArray(body)
+		}
+	}
 	return &msg, err
 }
 
+func unmarshalMessageArray(body string) (*jsonrpcMessage, error) {
+	var msgs []*jsonrpcMessage
+	err := json.Unmarshal([]byte(body), &msgs)
+	if err != nil {
+		return nil, err
+	}
+
+	// return first element
+	if len(msgs) == 0 {
```

Comment:
What if `len(msgs) > 1`?

Reply:
@adambabik  until we know the nature of this web3.js behaviour, let's assume we just ok to take first element.

Reply:
What about printing this case as `WARN: Unhandled web3 payload case: xxx`? This way we could notice this from status-react side too.

Reply:
But shouldn't we throw an error in such a case? Or at least log it so that it may help us to figure out the root cause.

Reply:
Yes, indeed. @divan you should return a slice by default and let the caller handle this part of checking if only one item exists. Same for line 118 returns type.

Reply:
@oskarth are you referring to the same case as @adambabik? When we have two or more objects in array?

I personally don't think this case is even possible, but I agree it's cheap to add more validations and don't trust web3.js on generated payload.

Reply:
@divan yes

Reply:
Added.